File: synara/src/app/features/matrix-dto/space.ts
/**
 * Space hierarchy summary DTO.
 */

import type { RoomId } from './ids';
import {
  hasForbiddenWireFields,
  isObject,
  optBoolean,
  optString,
  reqString,
  stringArray,
} from './parseUtil';

export type SpaceChild = {
  roomId: RoomId;
  order?: string;
  suggested?: boolean;
};

export type SpaceSummary = {
  roomId: RoomId;
  name?: string;
  avatarUrl?: string;
  children: SpaceChild[];
  parentRoomIds?: RoomId[];
};

function parseChild(value: unknown): SpaceChild | null {
  if (!isObject(value)) return null;
  const roomId = reqString(value, 'roomId');
  const order = optString(value, 'order');
  const suggested = optBoolean(value, 'suggested');
  if (roomId === null || order === null || suggested === null) return null;
  return { roomId, order, suggested };
}

export function parseSpaceSummary(value: unknown): SpaceSummary | null {
  if (!isObject(value) || hasForbiddenWireFields(value)) return null;
  const roomId = reqString(value, 'roomId');
  const name = optString(value, 'name');
  const avatarUrl = optString(value, 'avatarUrl');
  if (roomId === null || name === null || avatarUrl === null || !Array.isArray(value.children)) {
    return null;
  }
  const children: SpaceChild[] = [];
  for (const c of value.children) {
    const parsed = parseChild(c);
    if (!parsed) return null;
    children.push(parsed);
  }
  let parentRoomIds: string[] | undefined;
  if (value.parentRoomIds !== undefined) {
    const arr = stringArray(value.parentRoomIds);
    if (arr === null) return null;
    parentRoomIds = arr;
  }
  return { roomId, name, avatarUrl, children, parentRoomIds };
}

File: crates/synara-core/src/dto/space.rs
//! Space hierarchy summary DTO.

use serde::{Deserialize, Serialize};

use super::ids::RoomId;

/// Child room/space edge in a space hierarchy.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SpaceChild {
    pub room_id: RoomId,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub order: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub suggested: Option<bool>,
}

/// Space summary for lobby / hierarchy UI.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SpaceSummary {
    pub room_id: RoomId,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub avatar_url: Option<String>,
    pub children: Vec<SpaceChild>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub parent_room_ids: Option<Vec<RoomId>>,
}
